### PR TITLE
IPv6 support

### PR DESCRIPTION
**Pull request by majst01**
by using inet.af/netaddr for ip math, this also enables us to have variable child prefix length.

API Breakage:
- change prefix length to uint8 from int
- Usage.AvailablePrefixes changed to AvailableSmallestPrefixes

TODO:
- [x] write tests for ipv6 addresses and prefixes
- [x] decide howto cope with variable child prefix length and the calculation of availablePrefixes.
- [x] write integration test with existing production if behavior is the same 
- [x] remove same child prefix length restriction
- [x] decide what to do with Prefix.childPrefixLength on write to be able to roll back. Keep this field.

The current integration test is based on data from our testing environment, should be done for production data as well.
~For comparison #38 should be merged first.~

Benchmarks:
```
name                    old time/op    new time/op    delta
NewPrefixMemory-4         3.15µs ± 1%    2.34µs ± 7%   -25.56%  (p=0.016 n=4+5)
NewPrefixPostgres-4       10.3ms ± 6%    10.1ms ±12%      ~     (p=0.686 n=4+4)
NewPrefixCockroach-4      81.1ms ± 5%    80.0ms ± 3%      ~     (p=0.486 n=4+4)
AcquireIPMemory-4         4.63µs ±12%    3.90µs ± 1%   -15.91%  (p=0.016 n=5+4)
AcquireIPPostgres-4       12.6ms ± 9%    13.4ms ± 7%      ~     (p=0.222 n=5+5)
AcquireIPCockroach-4      79.6ms ± 6%    85.2ms ± 9%      ~     (p=0.095 n=5+5)
AcquireChildPrefix1-4     31.6µs ±12%     8.1µs ± 7%   -74.43%  (p=0.008 n=5+5)
AcquireChildPrefix2-4     93.8µs ±17%     8.3µs ±10%   -91.17%  (p=0.008 n=5+5)
AcquireChildPrefix3-4     1.56ms ± 7%    0.01ms ± 5%   -99.51%  (p=0.008 n=5+5)
AcquireChildPrefix4-4     9.79ms ±37%    0.01ms ± 4%   -99.92%  (p=0.008 n=5+5)
AcquireChildPrefix5-4     82.4ms ± 6%     0.0ms ±34%   -99.99%  (p=0.008 n=5+5)
AcquireChildPrefix6-4     10.9µs ±10%     7.9µs ± 5%   -27.17%  (p=0.008 n=5+5)
AcquireChildPrefix7-4     15.0µs ± 6%     7.7µs ± 2%   -48.61%  (p=0.008 n=5+5)
AcquireChildPrefix8-4     30.2µs ± 0%     7.5µs ± 1%   -75.04%  (p=0.016 n=4+5)
AcquireChildPrefix9-4     97.2µs ± 9%     7.7µs ± 3%   -92.10%  (p=0.008 n=5+5)
AcquireChildPrefix10-4     360µs ± 7%       8µs ± 1%   -97.91%  (p=0.008 n=5+5)
PrefixOverlapping-4       1.65µs ±12%    0.34µs ± 1%   -79.25%  (p=0.008 n=5+5)

name                    old alloc/op   new alloc/op   delta
NewPrefixMemory-4         1.80kB ± 0%    1.44kB ± 0%   -20.00%  (p=0.008 n=5+5)
NewPrefixPostgres-4       6.07kB ± 1%    5.82kB ± 1%    -4.10%  (p=0.029 n=4+4)
NewPrefixCockroach-4      7.63kB ± 6%    7.44kB ± 6%      ~     (p=0.686 n=4+4)
AcquireIPMemory-4         2.31kB ± 0%    2.25kB ± 0%    -2.77%  (p=0.008 n=5+5)
AcquireIPPostgres-4       10.8kB ± 1%    10.9kB ± 2%      ~     (p=0.056 n=5+5)
AcquireIPCockroach-4      12.7kB ± 2%    13.0kB ± 2%    +2.73%  (p=0.008 n=5+5)
AcquireChildPrefix1-4     15.5kB ± 0%     4.3kB ± 0%   -72.12%  (p=0.008 n=5+5)
AcquireChildPrefix2-4     48.3kB ± 0%     4.3kB ± 0%   -91.05%  (p=0.008 n=5+5)
AcquireChildPrefix3-4      699kB ± 0%       4kB ± 0%   -99.38%  (p=0.008 n=5+5)
AcquireChildPrefix4-4     2.80MB ± 0%    0.00MB ± 0%   -99.85%  (p=0.000 n=4+5)
AcquireChildPrefix5-4     14.9MB ± 4%     0.0MB ± 0%   -99.97%  (p=0.008 n=5+5)
AcquireChildPrefix6-4     5.24kB ± 0%    4.32kB ± 0%   -17.56%  (p=0.008 n=5+5)
AcquireChildPrefix7-4     7.18kB ± 0%    4.32kB ± 0%   -39.82%  (p=0.008 n=5+5)
AcquireChildPrefix8-4     15.5kB ± 0%     4.3kB ± 0%   -72.12%  (p=0.008 n=5+5)
AcquireChildPrefix9-4     48.3kB ± 0%     4.3kB ± 0%   -91.05%  (p=0.008 n=5+5)
AcquireChildPrefix10-4     202kB ± 0%       4kB ± 0%   -97.86%  (p=0.008 n=5+5)
PrefixOverlapping-4         432B ± 0%        0B       -100.00%  (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
NewPrefixMemory-4           41.0 ± 0%      20.0 ± 0%   -51.22%  (p=0.008 n=5+5)
NewPrefixPostgres-4          149 ± 1%       128 ± 1%   -13.78%  (p=0.029 n=4+4)
NewPrefixCockroach-4         166 ± 3%       144 ± 2%   -12.95%  (p=0.029 n=4+4)
AcquireIPMemory-4           49.0 ± 0%      42.0 ± 0%   -14.29%  (p=0.008 n=5+5)
AcquireIPPostgres-4          267 ± 0%       260 ± 0%    -2.70%  (p=0.016 n=4+5)
AcquireIPCockroach-4         292 ± 1%       286 ± 1%    -2.06%  (p=0.008 n=5+5)
AcquireChildPrefix1-4        113 ± 0%        69 ± 0%   -38.94%  (p=0.008 n=5+5)
AcquireChildPrefix2-4        113 ± 0%        69 ± 0%   -38.94%  (p=0.008 n=5+5)
AcquireChildPrefix3-4        326 ± 7%        69 ± 0%   -78.81%  (p=0.008 n=5+5)
AcquireChildPrefix4-4      5.20k ± 3%     0.07k ± 0%   -98.67%  (p=0.000 n=4+5)
AcquireChildPrefix5-4       161k ±14%        0k ± 0%   -99.96%  (p=0.008 n=5+5)
AcquireChildPrefix6-4        109 ± 0%        69 ± 0%   -36.70%  (p=0.008 n=5+5)
AcquireChildPrefix7-4        109 ± 0%        69 ± 0%   -36.70%  (p=0.008 n=5+5)
AcquireChildPrefix8-4        113 ± 0%        69 ± 0%   -38.94%  (p=0.008 n=5+5)
AcquireChildPrefix9-4        113 ± 0%        69 ± 0%   -38.94%  (p=0.008 n=5+5)
AcquireChildPrefix10-4       125 ± 0%        69 ± 0%   -44.62%  (p=0.008 n=5+5)
PrefixOverlapping-4         24.0 ± 0%       0.0       -100.00%  (p=0.008 n=5+5)
```
We now have a constant time for Prefix allocation, regardless of length, and smaller memory footprint and allocations.

@droid42 @mwindower @Gerrit91 this is still WIP, but would appreciate an early review. This would bring us a large step forward.

### CODE REVIEW

**Comment by majst01** [comment]
Raised a PR in https://github.com/inetaf/netaddr/pull/61 to move my childprefix logic upstream

**Comment by mwindower** [comment]
Very nice to see this progressing!

Here some proposals for aligning names a bit more to standard terms used in ip address/cidr calculation space (https://docs.python.org/3/library/ipaddress.html#networks-as-containers-of-addresses):
- extractPrefix              -> subnet
- extractPrefixFromSet -> nextFreeSubnet
- availablePrefixes       -> subnetCount
- childPrefixLength       -> subnetPrefixLength

**Comment by majst01** [comment]
Thanks for looking at my code.

regarding availablePrefixes/subnetCount calculation:
The algorithm of netaddr.IPSet.Prefixes will return a slice of Prefixes as packed as possible like so:

```go
func TestPrefix(t *testing.T) {
	prefix, _ := netaddr.ParseIPPrefix("192.168.0.0/22")
	remove, _ := netaddr.ParseIPPrefix("192.168.0.0/28")
	remove2, _ := netaddr.ParseIPPrefix("192.168.1.0/28")

	var ipset netaddr.IPSet

	ipset.AddPrefix(prefix)
	ipset.RemovePrefix(remove)
	ipset.RemovePrefix(remove2)
	for _, p := range ipset.Prefixes() {
		t.Logf("prefix:%s", p)
	}
}
    ip_test.go:35: prefix:192.168.0.16/28
    ip_test.go:35: prefix:192.168.0.32/27
    ip_test.go:35: prefix:192.168.0.64/26
    ip_test.go:35: prefix:192.168.0.128/25
    ip_test.go:35: prefix:192.168.1.16/28
    ip_test.go:35: prefix:192.168.1.32/27
    ip_test.go:35: prefix:192.168.1.64/26
    ip_test.go:35: prefix:192.168.1.128/25
    ip_test.go:35: prefix:192.168.2.0/23
```
So in this case subnetCount would be 9 which will suggest we are short in subnets, and after acquiring one more subnet, we will have eventually a higher subnetCount.

**Comment by mwindower** [comment]
We could return a slice from `availablePrefixes` containing a histogram of prefix length and how often they are available. The slice should be sorted by prefix length - to the end user we could show e.g. that a `/24` is the biggest availablePrefix. WDYT?

Like this:
```golang

type AvailablePrefix struct {
  PrefixLength uint8
  Count uint32
}

AvailablePrefixes []AvailablePrefix  
```

**Comment by majst01** [comment]
I think this is a reasonable approach, albeit its still difficult to have a simple understanding how much space is left from a ui perspektive.

File: prefix.go [hunk 1]
```diff
@@ -110,14 +108,18 @@ func copyMap(m map[string]bool) map[string]bool {
...
-	AcquiredPrefixes  uint64
+	// AvailableIPs the number of available IPs if this is not a parent prefix
+	AvailableIPs uint64
+	// AcquiredIPs the number of acquire IPs if this is not a parent prefix
+	AcquiredIPs uint64
+	// AvailableSmallestPrefixes is the count of available Prefixes with 2 countable Bits
+	AvailableSmallestPrefixes uint64
+	// AcquiredPrefixes the number of acquired prefixes if this is a parent prefix
+	AcquiredPrefixes uint64
 }
```
Comment (majst01) [inline prefix.go:119]: Add AvailableChildPrefixes

Comment (majst01) [reply prefix.go:121]: Done

**Comment by majst01** [comment]
Added the first gardener ipv6 compatibility test: https://github.com/gardener/gardener/pull/3391

**Review by Gerrit91 — APPROVED** [review]

File: prefix_test.go [hunk 2]
```diff
@@ -686,3 +1148,62 @@ func TestGob(t *testing.T) {
...
+			got, avpfxs := p.availablePrefixes()
+			for _, pfx := range avpfxs {
+				// Only logs if fails
+				ipprefix, err := netaddr.ParseIPPrefix(pfx)
+				require.NoError(t, err)
+				smallest := 1 << (ipprefix.IP.BitLen() - 2 - ipprefix.Bits)
+				t.Logf("available prefix:%s smallest left:%d", pfx, smallest)
+			}
+
+			if diff := deep.Equal(tt.want, got); diff != nil {
```
Comment (droid42) [inline prefix_test.go:1204]: You can simply write `if tt.want != got {` since both variables are uint64

**Review by droid42 — APPROVED** [review]
LGTM, came across only one small possible simplification

File: prefix_test.go (hunk 2 above)
Comment (majst01) [reply prefix_test.go:1204]: will fix

Comment (majst01) [reply prefix_test.go]: done

**Review by mwindower — APPROVED** [review]
lgtm

**Pull request merged**